gen3analysis/routes/survival.py

```diff
@@ -282,3 +286,115 @@ async def plot(
+    responses={
+        status.HTTP_200_OK: {"description": "Successfully processed the survival plot"},
+        status.HTTP_400_BAD_REQUEST: {
+            "description": "The request body is missing required fields or has invalid values."
+        },
+        status.HTTP_401_UNAUTHORIZED: {
+            "description": "User unauthorized when accessing endpoint"
+        },
+        status.HTTP_403_FORBIDDEN: {
+            "description": "User does not have access to requested data"
+        },
+        status.HTTP_500_INTERNAL_SERVER_ERROR: {
+            "description": "Something went wrong internally when processing the request"
+        },
+    },
+)
+async def compare(
+    request: CompareSurvivalRequest,
+    gen3_graphql_client: GuppyGQLClient = Depends(get_guppy_client),
+    auth: Auth = Depends(Auth),
+) -> JSONResponse:
+    filters = request.filters
+    doc_type = request.doc_type
+    field = request.field
+    limit = request.limit
+
+    ## get the cases
+
+    if len(filters) != 2:
+        raise HTTPException(
+            status_code=400, detail="filters must be a list of 2 filters"
+        )
+
+    if field is None:
+        raise HTTPException(status_code=400, detail="field must be specified")
+
+    if doc_type is None:
+        raise HTTPException(status_code=400, detail="doc_type must be specified")
+
+    # get a list of cases to perform set operation on, for each cohort
+    try:
+        plot_items_0 = await cases.get_item_ids(
+            gen3_graphql_client, auth, doc_type, field, filters[0], limit
+        )
+
+        plot_items_1 = await cases.get_item_ids(
+            gen3_graphql_client, auth, doc_type, field, filters[1], limit
+        )
+
+        if plot_items_0.get("data") is None:
+            raise HTTPException(
+                status_code=400, detail="No cases found for the first filter"
+            )
+        if plot_items_1.get("data") is None:
+            raise HTTPException(
+                status_code=400, detail="No cases found for the second filter"
+            )
+
+        # extract ids
+
+        ids = glom(plot_items_0, f"data.{doc_type}", default=[])
+        ids_0 = [x[field] for x in ids if x.get(field) is not None]
+
+        ids = glom(plot_items_1, f"data.{doc_type}", default=[])
+        ids_1 = [x[field] for x in ids if x.get(field) is not None]
+
+        # Fastest approach: Convert to sets once and reuse
+        set0 = set(ids_0)
+        set1 = set(ids_1)
+
+        intersection = set0 & set1
+
+        # Subtract the intersection from both sets
+        item_id_0_minus_intersection = list(set0 - intersection)
+        item_id_1_minus_intersection = list(set1 - intersection)
+
+        # build graphql filter for both using in
+
+        filter_0 = {"and": [{"in": {field: item_id_0_minus_intersection}}]}
+
+        filter_1 = {"and": [{"in": {field: item_id_1_minus_intersection}}]}
+
+        return await plot(
+            PlotRequest(filters=[filter_0, filter_1]),
```

Comment:
If i understand the logic correctly, this endpoint could be simpler (and likely more performant) if guppy supported one of these:
- "not": you could do this in a single query, and avoid manipulating lists of 10k IDs:
```
filter_0 = {"and": [{filters[0]}, {"not": filters[1]}]}
filter_1 = {"and": [{filters[1]}, {"not": filters[0]}]}
```
- "not in": you could do this instead of calculating the intersection:
```
filter_0 = {"and": [{filters[0]}, {"not in": {field: ids_1}}]}
filter_1 = {"and": [{filters[1]}, {"not in": {field: ids_0}}]}
```
Just throwing it out there, in case it would be easy to add to guppy...

Reply:
That would be better for sure. However, MMRF is doing UAT next week on Cohort Comparison, so this will have to be a TODO :) .